### PR TITLE
Created A CherryPick Tactic

### LINKED ISSUE #464: Implement moveToOffensivePosition Tactic
### Description of the task

This tactic should move the robot to a generic offensive position on the field.

### Outline the task in more detail

An offensive position is considered somewhere on the field close to the enemy side, and in relatively open space. 2 possible implementations:
* Move the offensive robot to the "most open area" on the field, preferably near the enemy half. This could use the "find open space in triangles" function from the indirect chip and chase evaluations.
* Move the offensive robot to the receiver location of the best pass, assuming the bass originates at the ball positions. The region could be bounded to the enemy half of the field or similar. This should also choose relatively open areas on the field that would be a good place to be if we suddenly got possession of the ball.

### Acceptance criteria

- [ ] Implement the tactic
- [ ] unit tests
- [ ] sanity check in grsim

### PR DESCRIPTION
**Pull request by garethellis0**
### Description
This tactic is intended to place a robot in a given region, and have the robot constantly move towards the best location within that region such that if we were to pass the ball from it's current position, the robot would be in the best possible place to receive it.

### Testing Done
Some unit testing for geom additions (still broken until I get a quick clarification). There are no tests for the tactic though. The main issue is that most of the positioning work is done by the pass generator, so there really isn't any logic in the tactic itself. I'm totally open if anyone has suggestions for more tests to add though!

### Resolved Issues
resolves #464 

### Length Justification
NA

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/thunderbots/software/ai/passing/pass.cpp [hunk 1]
```diff
@@ -24,6 +26,11 @@ Point Pass::receiverPoint() const
     return receiver_point;
 }
 
+Angle Pass::receiverAngle() const
+{
+    return vertexAngle({0, 0}, receiverPoint(), passerPoint());
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/passing/pass.cpp]: If this is supposed to return the angle/orientation the receiver should face, you can do `(passerPoint() - receiverPoint()).orientation()`

File: src/thunderbots/software/ai/passing/pass.h [hunk 2]
```diff
@@ -39,6 +39,13 @@ namespace AI::Passing
          */
         Point receiverPoint() const;
 
+        /**
+         * Gets the angle the receiver should be facing to receive the pass
+         *
+         * @return The angle the receiver should be facing to receive the pass
+         */
+        Angle receiverAngle() const;
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/passing/pass.h]: `receiverOrientation` might be slightly more descriptive here.

File: src/thunderbots/software/ai/hl/stp/tactic/cherry_pick_tactic.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+std::unique_ptr<Intent> CherryPickTactic::calculateNextIntent(
+    intent_coroutine::push_type& yield)
+{
+    MoveAction move_action = MoveAction();
+    do
+    {
+        // Move the robot to be the best possible receiver for the best pass we can
+        // find (within the target region)
+        std::optional<AI::Passing::Pass> best_pass = pass_generator.getBestPassSoFar();
+        if (best_pass)
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/hl/stp/tactic/cherry_pick_tactic.cpp:42]: Would this `best_pass` ever be a `nullopt`? If `best_pass` is `nullopt` this function will get stuck in an infinite loop. Is there a reasonable fallback if `best_pass` resolves to `false`? Maybe just use the previous position?

Also, we should probably add hysteresis to the position being moved to.

File: src/thunderbots/software/ai/hl/stp/tactic/cherry_pick_tactic.cpp [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            yield(move_action.updateStateAndGetNextIntent(
+                *robot, best_pass->receiverPoint(), best_pass->receiverAngle(), 0));
+            best_pass->passerPoint();
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/hl/stp/tactic/cherry_pick_tactic.cpp]: What's this line doing?

File: src/thunderbots/software/ai/hl/stp/tactic/block_shot_path_tactic.h [hunk 5]
```diff
@@ -17,7 +17,8 @@ class BlockShotPathTactic : public Tactic
      * @param loop_forever Whether or not this Tactic should never complete. If true, the
      * tactic will be restarted every time it completes
      */
-    explicit BlockShotPathTactic(const Field& field, bool loop_forever = false);
+    // TODO: Should this loop forever???
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/hl/stp/tactic/block_shot_path_tactic.h]: I think this is fine to have loop_forever by default, or you can remove the option entirely if it should _always_ run forever. In this case, I think that's fine just make sure to mention it in the javadoc comment.

In the future I'm considering not giving the `loop_forever` a default value so it's very explicit which it's doing. That way we avoid ambiguous usage in the `Play`

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: src/thunderbots/software/ai/hl/stp/tactic/block_shot_path_tactic.h (hunk 5 above)
Comment (EvanMorcom) [reply src/thunderbots/software/ai/hl/stp/tactic/block_shot_path_tactic.h]: I'm all for removing the default value

File: src/thunderbots/software/ai/hl/stp/tactic/cherry_pick_tactic.h [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+   private:
+    std::unique_ptr<Intent> calculateNextIntent(
+        intent_coroutine::push_type& yield) override;
+
+    /**
+     * Calculates the location to move to in order to block the shot.
+     * @return the Point to move to in order to block the shot
+     */
+    Point getBlockPosition();
```
Comment (EvanMorcom) [inline src/thunderbots/software/ai/hl/stp/tactic/cherry_pick_tactic.h]: Is this function for a friendly robot to block the path towards the pass receiver?

File: src/thunderbots/software/ai/hl/stp/tactic/block_shot_path_tactic.h (hunk 5 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/hl/stp/tactic/block_shot_path_tactic.h]: Sorry, didn't mean to touch this file. Removed all changes to it.

Comment (garethellis0) [reply src/thunderbots/software/ai/hl/stp/tactic/block_shot_path_tactic.h]: Removed the default value in `cherry_pick_tactic.h`

File: src/thunderbots/software/ai/hl/stp/tactic/cherry_pick_tactic.cpp (hunk 4 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/hl/stp/tactic/cherry_pick_tactic.cpp]: nothing, removed

File: src/thunderbots/software/ai/hl/stp/tactic/cherry_pick_tactic.cpp (hunk 3 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/hl/stp/tactic/cherry_pick_tactic.cpp:42]: Theoretically it shouldn't, but you're totally right that we should deal with case explicitly anyhow. Changed the `while` conditional to 
```
    } while (!move_action.done() && best_pass);
```

With regards to hysteresis, are there some other examples I could take a look at of how we've done it, so we're being consistent? Or should I just come up with a semi-general scheme?

File: src/thunderbots/software/ai/hl/stp/tactic/cherry_pick_tactic.h (hunk 6 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/hl/stp/tactic/cherry_pick_tactic.h]: Whoops, copy-pasta leftover. Removed

File: src/thunderbots/software/ai/passing/pass.cpp (hunk 1 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/pass.cpp]: much better, thanks. changed

File: src/thunderbots/software/ai/passing/pass.h (hunk 2 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/pass.h]: renamed

File: src/thunderbots/software/ai/hl/stp/tactic/cherry_pick_tactic.cpp (hunk 3 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/tactic/cherry_pick_tactic.cpp:42]: semi-general is fine. We don't have anything like this that I'm aware of right now.

**Review by MathewMacDougall — COMMENTED** [review]
LGTM pending hysteresis and CI

File: src/thunderbots/software/ai/hl/stp/tactic/cherry_pick_tactic.cpp (hunk 3 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/hl/stp/tactic/cherry_pick_tactic.cpp:42]: After discussion on slack, decided to follow the principle of [YAGNI](https://www.martinfowler.com/bliki/Yagni.html) and leave this out until there is demonstrable need for it.

**Review by MathewMacDougall — APPROVED** [review]
LGTM :+1:

**Pull request merged**